{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook can be used to generate fake structural variant test data for testing the genome finishing module.\n",
    "Given a source fasta, bam, and paired-end fastqs and insertion parameters, it creates a directory with the following files:\n",
    "\n",
    "    ref.fa\n",
    "    reads.1.fq\n",
    "    reads.2.fq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/wahern/projects/millstone/genome_designer'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "from django.core.management import setup_environ\n",
    "import settings\n",
    "\n",
    "setup_environ(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import random\n",
    "import re\n",
    "import os\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pysam\n",
    "\n",
    "from genome_finish.millstone_de_novo_fns import get_avg_genome_coverage\n",
    "\n",
    "\n",
    "# def _make_fake_insertion(ref_endpoints, ins_endpoints):\n",
    "\n",
    "ref_endpoints = (2930000, 2940000)\n",
    "ins_endpoints = (2932000, 2933000)\n",
    "desired_coverage = 40\n",
    "output_no_insertion_ref = True\n",
    "\n",
    "test_number = 6\n",
    "\n",
    "\n",
    "template_dir = '/home/wahern/projects/millstone/genome_designer/test_data/genome_finish_test/mg1655_test/templates'\n",
    "\n",
    "source_fasta = os.path.join(template_dir, 'mg1655.fa')\n",
    "source_bam = os.path.join(template_dir, 'lib1_rec07.bwa_align.bam')\n",
    "source_fq1 = os.path.join(template_dir, 'lib1_rec07.1.fq')\n",
    "source_fq2 = os.path.join(template_dir, 'lib1_rec07.2.fq')\n",
    "\n",
    "\n",
    "test_dir = '/home/wahern/projects/millstone/genome_designer/test_data/genome_finish_test/mg1655_test'\n",
    "output_dir = os.path.join(test_dir, str(test_number))\n",
    "\n",
    "output_fasta = os.path.join(output_dir, 'ref.fa')\n",
    "output_fq1 = os.path.join(output_dir, 'reads.1.fq')\n",
    "output_fq2 = os.path.join(output_dir, 'reads.2.fq')\n",
    "\n",
    "assert not os.path.exists(output_dir)\n",
    "\n",
    "# Get sample\n",
    "if desired_coverage:\n",
    "    coverage = get_avg_genome_coverage(source_bam)\n",
    "    if desired_coverage > coverage:\n",
    "        raise Exception('Desired coverage:' + str(desired_coverage) +\n",
    "                ' is greater than the genome\\'s average coverage of ' +\n",
    "                str(coverage))\n",
    "        \n",
    "    read_sampling_fraction = desired_coverage / coverage\n",
    "    \n",
    "    linecount = 0\n",
    "    with open(source_fq1) as fh:\n",
    "        for line in fh:\n",
    "            linecount+=1\n",
    "    \n",
    "    include_read = [random.random() < read_sampling_fraction for i in xrange(linecount)]\n",
    "    \n",
    "    \n",
    "\n",
    "os.mkdir(output_dir)\n",
    "\n",
    "# Get source seqrecord\n",
    "with open(source_fasta, 'r') as source_fasta_fh:\n",
    "    source_seqrecord = SeqIO.parse(source_fasta_fh, 'fasta').next()\n",
    "    output_seqrecord = (\n",
    "            source_seqrecord[ref_endpoints[0]:ins_endpoints[0]] +\n",
    "            source_seqrecord[ins_endpoints[1]:ref_endpoints[1]]\n",
    "    )\n",
    "\n",
    "# Sanity check\n",
    "assert len(output_seqrecord) == ref_endpoints[1] - ref_endpoints[0] - (\n",
    "        ins_endpoints[1] - ins_endpoints[0])\n",
    "\n",
    "# Add some metadata to the header.\n",
    "output_seqrecord.id = source_seqrecord.id\n",
    "output_seqrecord.description = source_seqrecord.description + (\n",
    "        ', FAKE_SHORT: ' +\n",
    "        '{ref_endpoints:' + str(ref_endpoints) + ', ' +\n",
    "        'ins_endpoints:' + str(ins_endpoints) + '}')\n",
    "\n",
    "# Write output fasta\n",
    "with open(output_fasta, 'w') as fh:\n",
    "    SeqIO.write([output_seqrecord], fh, 'fasta')\n",
    "\n",
    "# Get reads in region\n",
    "qnames_in_region = {}\n",
    "source_af = pysam.AlignmentFile(source_bam)\n",
    "for read in source_af.fetch('NC_000913', ref_endpoints[0], ref_endpoints[1]):\n",
    "    if (not read.is_unmapped and\n",
    "            ref_endpoints[0] <= read.reference_start <= ref_endpoints[1]  and\n",
    "            ref_endpoints[0] <= read.reference_end <= ref_endpoints[1]):\n",
    "        qnames_in_region[read.qname] = True\n",
    "source_af.close()\n",
    "\n",
    "# Go through fastqs and write reads in ROI to file\n",
    "p1 = re.compile('@(\\S+)')\n",
    "for input_fq_path, output_fq_path in [(source_fq1, output_fq1), (source_fq2, output_fq2)]:\n",
    "    counter = 0\n",
    "    if desired_coverage:\n",
    "        iterator = iter(include_read)\n",
    "    with open(input_fq_path, 'r') as in_fh, open(output_fq_path, 'w') as out_fh:\n",
    "        for line in in_fh:\n",
    "            m1 = p1.match(line)\n",
    "            if m1:\n",
    "                qname = m1.group(1)\n",
    "                if qname in qnames_in_region:\n",
    "                    if desired_coverage:\n",
    "                        if iterator.next():\n",
    "                            out_fh.write(line)\n",
    "                            out_fh.write(in_fh.next())\n",
    "                            out_fh.write(in_fh.next())\n",
    "                            out_fh.write(in_fh.next())\n",
    "                    else:\n",
    "                        out_fh.write(line)\n",
    "                        out_fh.write(in_fh.next())\n",
    "                        out_fh.write(in_fh.next())\n",
    "                        out_fh.write(in_fh.next())\n",
    "                        \n",
    "\n",
    "if output_no_insertion_ref:\n",
    "    # Get source seqrecord\n",
    "    with open(source_fasta, 'r') as source_fasta_fh:\n",
    "        source_seqrecord = SeqIO.parse(source_fasta_fh, 'fasta').next()\n",
    "        output_seqrecord = (\n",
    "                source_seqrecord[ref_endpoints[0]:ref_endpoints[1]]\n",
    "        )\n",
    "\n",
    "    # Add some metadata to the header.\n",
    "    output_seqrecord.id = source_seqrecord.id\n",
    "    output_seqrecord.description = source_seqrecord.description + (\n",
    "            ', FAKE_SHORT: ' +\n",
    "            '{ref_endpoints:' + str(ref_endpoints) + ', ' +\n",
    "            'ins_endpoints:None}')\n",
    "\n",
    "    # Write output fasta\n",
    "    no_ins_fasta = os.path.join(output_dir, 'no_ins_ref.fa')\n",
    "    with open(no_ins_fasta, 'w') as fh:\n",
    "        SeqIO.write([output_seqrecord], fh, 'fasta')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
